### PR TITLE
Comment parsing bug when there is a line with nothing but comments

### LINKED ISSUE #381: Skip commented lines
I expected the `read.csv()` behaviour (TBH I'm surprised about the lack of `NA`s there).

``` r
library(readr)

mat <- matrix(
  c(paste0("thing", 1:3),
    "A2", "B2", "C2",
    "A3", "#B3", "C3",
    "#A4", "B4", "C4",
    "A5", "*", "C5"),
  byrow = TRUE, ncol = 3)
row_txt <- apply(mat, 1, paste, collapse = ",")
txt <- paste(row_txt, collapse = "\n")

read_csv(txt)
#> Source: local data frame [4 x 3]
#> 
#>   thing1 thing2 thing3
#>    (chr)  (chr)  (chr)
#> 1     A2     B2     C2
#> 2     A3    #B3     C3
#> 3    #A4     B4     C4
#> 4     A5      *     C5
read.csv(text = txt)
#>   thing1 thing2 thing3
#> 1     A2     B2     C2
#> 2     A3    #B3     C3
#> 3    #A4     B4     C4
#> 4     A5      *     C5

read_csv(txt, comment = "#")
#> Warning: 1 parsing failure.
#> row col  expected    actual
#>   2  -- 3 columns 4 columns
#> Source: local data frame [2 x 3]
#> 
#>   thing1 thing2 thing3
#>    (chr)  (chr)  (chr)
#> 1     A2     B2     C2
#> 2     A3     A5      *
read.csv(text = txt, comment.char = "#")
#>   thing1 thing2 thing3
#> 1     A2     B2     C2
#> 2     A3              
#> 3     A5      *     C5
```

### PR DESCRIPTION
**Pull request by jimhester**
Fixes #381

### CODE REVIEW

File: src/TokenizerDelim.cpp [hunk 1]
```diff
@@ -160,9 +161,17 @@ Token TokenizerDelim::nextToken() {
 
     case STATE_COMMENT:
       if (*cur_ == '\r' || *cur_ == '\n') {
+
+        // If we have read at least one record on the current row go to the
+        // next row, line, otherwise just ignore the line.
+        if (col_ > 0) {
+          newRecord();
+          ++row;
```
Comment (hadley) [inline src/TokenizerDelim.cpp]: I think we always need to increment `row` otherwise the error messages will be off after lines that contain only coments?

Comment (jimhester) [inline src/TokenizerDelim.cpp]: That is true, but then you get all NAs for commented out lines in the output, which differs from the current convention (and that of `read.csv()`).

Comment (hadley) [inline src/TokenizerDelim.cpp]: So maybe we need a second parameter? Lines vs. rows?

Comment (jimhester) [inline src/TokenizerDelim.cpp]: Yeah maybe that would work best.

Comment (hadley) [inline src/TokenizerDelim.cpp]: I'm having vague flashbacks that this is why I never fixed this earlier ;) Probaby interacts with skip too

**Comment by hadley** [comment]
Probably same problem in #321

**Comment by jimhester** [comment]
An alternative to tracking rows and lines separately would be to output a row of NAs for every blank line (including commented out lines). Benefits would be simple implementation and consistency between input and output numbers and I imagine there are times where users would want to keep blank lines. Downside is `read.csv()` drops blanks by default.

**Comment by hadley** [comment]
I think we really do need to drop missing lines by default

**Comment by jimhester** [comment]
Ok I will work on it, do we want an option to keep them like `read.csv()`?

**Comment by hadley** [comment]
It doesn't seem that useful to me, so I'd say no (we can always add later if people really want it)

**Pull request merged**